### PR TITLE
SLVS-938 Support object equality method calls

### PR DESCRIPTION
**Pull request by tvsonar**

### CODE REVIEW

File: src/SonarLint.CSharp/Helpers/SymbolicExecution/ExplodedGraph.cs [hunk 1]
```diff
@@ -243,8 +243,9 @@ protected override void VisitInstruction(ExplodedGraphNode node)
                             !check.TryProcessInstruction((MemberAccessExpressionSyntax)instruction, newProgramState, out newProgramState))
                         {
                             // Default behavior
-                            newProgramState = newProgramState.PopValue();
-                            newProgramState = newProgramState.PushValue(new SymbolicValue());
+                            SymbolicValue @object;
```
Comment (m-g-sonar) [inline src/SonarLint.CSharp/Helpers/SymbolicExecution/ExplodedGraph.cs]: That's the worst name ever. It's the SV associated to the member select expression, use a meaningful name!

File: src/SonarLint/Helpers/SymbolicExecution/SymbolicValues/MemberAccessSymbolicValue.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+
+namespace SonarLint.Helpers.FlowAnalysis.Common
+{
+    public class MemberAccessSymbolicValue : SymbolicValue
+    {
+        public SymbolicValue Object { get; }
+
+        public MemberAccessSymbolicValue(SymbolicValue @object)
```
Comment (m-g-sonar) [inline src/SonarLint/Helpers/SymbolicExecution/SymbolicValues/MemberAccessSymbolicValue.cs]: Please use `expression` or `memberExpression`, or anything your want, but something more meaningful than `@object`

File: src/SonarLint.CSharp/Helpers/SymbolicExecution/ExplodedGraph.cs [hunk 3]
```diff
@@ -516,6 +516,100 @@ private static ProgramState VisitBinaryOperator(ProgramState programState, Func<
...
+            SymbolicValue arg2;
+            var memberAccess = expression as MemberAccessSymbolicValue;
+            if (memberAccess != null)
+            {
+                arg2 = memberAccess.Object;
+            }
+            else
+            {
+                // this reference:
+                arg2 = new SymbolicValue();
```
Comment (m-g-sonar) [inline src/SonarLint.CSharp/Helpers/SymbolicExecution/ExplodedGraph.cs]: maybe worth mentioning the associated ticket

**Comment by tvsonar** [comment]
There's a flaw in the implementation. The assumption that Equals and ReferenceEquals are the same is incorrect: ReferenceEquals implies Equals, but not the other way.
This is a real problem in case of value types, which are never reference equal, but they can be equal.

File: src/SonarLint.CSharp/Helpers/SymbolicExecution/ExplodedGraph.cs [hunk 4]
```diff
@@ -516,6 +516,101 @@ private static ProgramState VisitBinaryOperator(ProgramState programState, Func<
...
+
+                baseMethod = baseMethod.OverriddenMethod;
+            }
+
+            return false;
+        }
+
+        private bool IsObjectEqualsCall(IMethodSymbol methodSymbol)
+        {
+            return methodSymbol != null &&
```
Comment (SonarTech) [inline src/SonarLint.CSharp/Helpers/SymbolicExecution/ExplodedGraph.cs]: MAJOR The 'this' parameter (or 'Me' in Visual Basic) of 'ExplodedGraph.IsObjectEqualsCall(IMethodSymbol)' is never used. Mark the member as static (or Shared in Visual Basic) or use 'this'/'Me' in the method body or at least one property accessor, if appropriate. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AMarkMembersAsStatic)

File: src/SonarLint.CSharp/Helpers/SymbolicExecution/ExplodedGraph.cs [hunk 5]
```diff
@@ -516,6 +516,101 @@ private static ProgramState VisitBinaryOperator(ProgramState programState, Func<
...
+                arg2 = new SymbolicValue();
+                newProgramState = arg2.SetConstraint(ObjectConstraint.NotNull, newProgramState);
+            }
+
+            return newProgramState.PushValue(new EqualsSymbolicValue(arg1, arg2));
+        }
+
+        private bool IsReferenceEqualsCall(IMethodSymbol methodSymbol)
+        {
+            return methodSymbol != null &&
```
Comment (SonarTech) [inline src/SonarLint.CSharp/Helpers/SymbolicExecution/ExplodedGraph.cs]: MAJOR The 'this' parameter (or 'Me' in Visual Basic) of 'ExplodedGraph.IsReferenceEqualsCall(IMethodSymbol)' is never used. Mark the member as static (or Shared in Visual Basic) or use 'this'/'Me' in the method body or at least one property accessor, if appropriate. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AMarkMembersAsStatic)

File: src/SonarLint.CSharp/Helpers/SymbolicExecution/ExplodedGraph.cs [hunk 6]
```diff
@@ -516,6 +516,101 @@ private static ProgramState VisitBinaryOperator(ProgramState programState, Func<
...
+                methodSymbol.ContainingType.Is(KnownType.System_Object) &&
+                methodSymbol.Name == "ReferenceEquals";
+        }
+
+        private bool IsInstanceEqualsCall(IMethodSymbol methodSymbol)
+        {
+            if (methodSymbol == null ||
```
Comment (SonarTech) [inline src/SonarLint.CSharp/Helpers/SymbolicExecution/ExplodedGraph.cs]: MAJOR The 'this' parameter (or 'Me' in Visual Basic) of 'ExplodedGraph.IsInstanceEqualsCall(IMethodSymbol)' is never used. Mark the member as static (or Shared in Visual Basic) or use 'this'/'Me' in the method body or at least one property accessor, if appropriate. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AMarkMembersAsStatic)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* MAJOR 5 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:D1AE804F-AE78-4883-B3D6-9E3C4026DEF6: Since 'EqualsRelationship' redefines Equals, it should also redefine GetHashCode. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AOverrideGetHashCodeOnOverridingEquals)
1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:D1AE804F-AE78-4883-B3D6-9E3C4026DEF6: Since 'NotEqualsRelationship' redefines Equals, it should also redefine GetHashCode. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AOverrideGetHashCodeOnOverridingEquals)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* MAJOR 3 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:D1AE804F-AE78-4883-B3D6-9E3C4026DEF6: Since 'EqualsRelationship' redefines Equals, it should also redefine GetHashCode. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AOverrideGetHashCodeOnOverridingEquals)
1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:D1AE804F-AE78-4883-B3D6-9E3C4026DEF6: Since 'NotEqualsRelationship' redefines Equals, it should also redefine GetHashCode. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AOverrideGetHashCodeOnOverridingEquals)

File: src/SonarLint/Helpers/SymbolicExecution/Relationships/EqualsRelationship.cs [hunk 7]
```diff
@@ -18,41 +18,16 @@
...
-                return false;
-            }
-
-            return Equals(obj as EqualsRelationship);
-        }
-
-        private bool Equals(EqualsRelationship other)
+        protected bool Equals(EqualsRelationship other)
         {
             return other != null && OperandsMatch(other);
```
Comment (SonarTech) [inline src/SonarLint/Helpers/SymbolicExecution/Relationships/EqualsRelationship.cs]: MAJOR Consider changing the type of parameter 'other' in 'EqualsRelationship.Equals(EqualsRelationship)' from 'EqualsRelationship' to its base type 'BinaryRelationship'. This method appears to only require base class members in its implementation. Suppress this violation if there is a compelling reason to require the more derived type in the method signature. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AConsiderPassingBaseTypesAsParameters)

File: src/SonarLint/Helpers/SymbolicExecution/Relationships/NotEqualsRelationship.cs [hunk 8]
```diff
@@ -18,43 +18,16 @@
...
-                return false;
-            }
-
-            return Equals(obj as NotEqualsRelationship);
-        }
-
-        private bool Equals(NotEqualsRelationship other)
+        protected bool Equals(NotEqualsRelationship other)
         {
             return other != null && OperandsMatch(other);
```
Comment (SonarTech) [inline src/SonarLint/Helpers/SymbolicExecution/Relationships/NotEqualsRelationship.cs]: MAJOR Consider changing the type of parameter 'other' in 'NotEqualsRelationship.Equals(NotEqualsRelationship)' from 'NotEqualsRelationship' to its base type 'BinaryRelationship'. This method appears to only require base class members in its implementation. Suppress this violation if there is a compelling reason to require the more derived type in the method signature. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AConsiderPassingBaseTypesAsParameters)

**Comment by SonarTech** [comment]
SonarQube analysis reported 8 issues
* MAJOR 8 major

Watch the comments in this conversation to review them.

#### 5 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:D1AE804F-AE78-4883-B3D6-9E3C4026DEF6: Change the accessibility of all public constructors in 'EqualsRelationship' to protected. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AAbstractTypesShouldNotHaveConstructors)
1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:D1AE804F-AE78-4883-B3D6-9E3C4026DEF6: Change the accessibility of all public constructors in 'NotEqualsRelationship' to protected. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AAbstractTypesShouldNotHaveConstructors)
1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:D1AE804F-AE78-4883-B3D6-9E3C4026DEF6: Change the accessibility of all public constructors in 'NotEqualsSymbolicValue' to protected. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AAbstractTypesShouldNotHaveConstructors)
1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:D1AE804F-AE78-4883-B3D6-9E3C4026DEF6: Since 'ValueEqualsRelationship' redefines Equals, it should also redefine GetHashCode. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AOverrideGetHashCodeOnOverridingEquals)
1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:D1AE804F-AE78-4883-B3D6-9E3C4026DEF6: Since 'ValueNotEqualsRelationship' redefines Equals, it should also redefine GetHashCode. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AOverrideGetHashCodeOnOverridingEquals)

File: src/SonarLint/Helpers/SymbolicExecution/Relationships/EqualsRelationship.cs [hunk 9]
```diff
@@ -18,43 +18,40 @@
...
-            return relationships
-                .OfType<NotEqualsRelationship>()
-                .Any(rel => OperandsMatch(rel));
-        }
-
-        public override string ToString()
+        protected bool Equals(EqualsRelationship other)
         {
-            return $"{LeftOperand} == {RightOperand}";
+            return other != null && OperandsMatch(other);
```
Comment (SonarTech) [inline src/SonarLint/Helpers/SymbolicExecution/Relationships/EqualsRelationship.cs]: MAJOR Consider changing the type of parameter 'other' in 'EqualsRelationship.Equals(EqualsRelationship)' from 'EqualsRelationship' to its base type 'BinaryRelationship'. This method appears to only require base class members in its implementation. Suppress this violation if there is a compelling reason to require the more derived type in the method signature. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AConsiderPassingBaseTypesAsParameters)

File: src/SonarLint/Helpers/SymbolicExecution/Relationships/NotEqualsRelationship.cs [hunk 10]
```diff
@@ -18,45 +18,29 @@
...
         {
         }
 
-        internal override bool IsContradicting(IEnumerable<BinaryRelationship> relationships)
+        protected bool Equals(NotEqualsRelationship other)
         {
-            return relationships
-                .OfType<EqualsRelationship>()
-                .Any(rel => OperandsMatch(rel));
+            return other != null && OperandsMatch(other);
```
Comment (SonarTech) [inline src/SonarLint/Helpers/SymbolicExecution/Relationships/NotEqualsRelationship.cs]: MAJOR Consider changing the type of parameter 'other' in 'NotEqualsRelationship.Equals(NotEqualsRelationship)' from 'NotEqualsRelationship' to its base type 'BinaryRelationship'. This method appears to only require base class members in its implementation. Suppress this violation if there is a compelling reason to require the more derived type in the method signature. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AConsiderPassingBaseTypesAsParameters)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* MAJOR 4 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:D1AE804F-AE78-4883-B3D6-9E3C4026DEF6: Since 'ValueEqualsRelationship' redefines Equals, it should also redefine GetHashCode. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AOverrideGetHashCodeOnOverridingEquals)
1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:D1AE804F-AE78-4883-B3D6-9E3C4026DEF6: Since 'ValueNotEqualsRelationship' redefines Equals, it should also redefine GetHashCode. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AOverrideGetHashCodeOnOverridingEquals)

**Pull request merged**